src/Compilers/CSharp/Portable/Compilation/SyntaxAndDeclarationManager.cs

```diff
@@ -565,6 +593,45 @@ public SyntaxAndDeclarationManager ReplaceSyntaxTree(SyntaxTree oldTree, SyntaxT
                 this.MessageProvider,
                 this.IsSubmission,
                 state);
+
+            OneOrMany<ImmutableSegmentedHashSet<string>> tryGetLastComputedMemberNames()
+            {
+                var previousRootNamespaceDeclaration = declMapBuilder[oldTree];
+                if (previousRootNamespaceDeclaration.IsValueCreated)
+                {
+                    // we computed the last root.  It will have up to date member names.  So just return those if present.
+                    var stack = s_declarationStack.Allocate();
+                    stack.Push(previousRootNamespaceDeclaration.Value);
+
+                    var builder = ArrayBuilder<ImmutableSegmentedHashSet<string>>.GetInstance();
+
+                    while (stack.Count > 0)
+                    {
+                        var current = stack.Pop();
+
+                        // As we walk down, push children on in reverse order so we get a DFS walk.
+
+                        for (int i = current.Children.Length - 1; i >= 0; i--)
+                            stack.Push(current.Children[i]);
+
+                        // Process any type we see before we process its nested types.
+                        if (current is SingleTypeDeclaration singleType && DeclarationTreeBuilder.CachesComputedMemberNames(singleType))
+                            builder.Add(singleType.MemberNames);
+                    }
+
+                    s_declarationStack.Free(stack);
+
+                    return builder.ToOneOrManyAndFree();
```

Comment:
>return builder.ToOneOrManyAndFree(); [](http://example.com/codeflow?start=20&length=36)

At the moment, it is not obvious to me that it is Ok to return a result that might be missing some entries due to ```CachesComputedMemberNames``` check.

Reply:
Added comments indicating why this is ok.